docs/proposals/event-notification.md

```diff
@@ -0,0 +1,143 @@
+{ "devFileActionExecutionBegun" : { "commandName" : "build", "actionName" : "run-build-script", "timestamp" : "(...)" } }
+{ "logText" : { "text:" "one line of text received\\n another line of text received", "timestamp" : "(...)" } }
+{ "devFileActionExecutionComplete" : { "errorCode" : 1, ( same as above )} }
+{ "logText" : { "text": (... ), "timestamp" : "(...)" } } # Log text is interleaved with events 
+{ "devFileCommandExecutionComplete": { "success" : false, (same as above) } }
+
+(Exact details on event name, and JSON format are TBD; feedback welcome!)
+```
+
+These events allow an external odo-consuming tool to determine the build status of an application (build succeeded, build failed, build not running).
+
+Note that unlike other machine-readable outputs used in odo, each individual line is a fully complete and parseable JSON document, allowing events to be streamed and processed by the consuming tool one-at-a-time, rather than waiting for all the events to be received before being parseable (which would be required if the entire console output was one single JSON document, as is the case for other odo machine-readable outputs.)
+
+### Detailed build status via JSON+custom markup
+
+For detailed build status, it is proposed that devfile writers may *optionally* include custom markup in their devfile actions which indicate a detailed build status:
+- If a dev file writer wanted to communicate that the current command/action were compiling the application, they would insert a specific markup string (`#devfile-status#`) at the beginning of a console-outputted line, and then between those two fields would be a JSON object with a single field `buildStatus`:
+  - For example: `#devfile-status# {"buildStatus":"Compiling application"}` would then communicate that the detailed build status should be set to `Compiling application`.
+- Since this line would be output as container stdout, it would be included as a `logText` JSON event, and the consuming tool can look for this markup string and parse the simple JSON to extract the detailed build status.
+- Feedback welcome around exact markup text format.
+
+The build step (running as a bash script, for example, invoked via an action) of a devfile might then look like this:
+```
+#!/bin/bash
+(...)
+echo "#devfile-status# {'buildStatus':'Compiling application'}
+mvn compile
+echo "#devfile-status# {'buildStatus':'Running unit tests'}
+mvn test
+```
+
+This 'detailed build status' markup text is *entirely optional*: if this markup is not present, the odo tool can still determine build succeeded/failed and build running/not-running using the other `odo push -o json` JSON events. 
+
+### App status notification via `odo component status -o json` and `odo log --follow`
+
+In general, within the execution context that odo operates, there are a few ways for us to determine the application status:
+1) Can the application be pinged at its exposed URL? 
```

Comment:
Good catch: Every case where I wrote 'ping' in the document, I actually meant 'send an HTTP/S request'... since 'ping' is is ambiguous language for this idea I've removed the term and replaced it with better HTTP/S-specific language.